File: src/oop/Object/Legacy_StringObjectContainerEntry.c
#include "Legacy_StringObjectContainerEntry.r"

/**
 * @brief `Destructs`s the given legacy_stringObjectContainerEntry, and returns its
 *        `legacy_stringObjectContainerEntry->value`.
 *
 * @param stringObjectContainerEntry
 * @return
 */
void *Legacy_StringObjectContainerEntryDestructor(
        Legacy_StringObjectContainerEntry *stringObjectContainerEntry) {
    if (stringObjectContainerEntry == NULL) { return NULL; }

    void *valueDestructorReturnValue =
            stringObjectContainerEntry->value->legacyObjectComponent
                    ->destructable->destructor(
                    stringObjectContainerEntry->value);

    free(stringObjectContainerEntry->object);

    free(stringObjectContainerEntry);

    return valueDestructorReturnValue;
}

void constructor_Legacy_StringObjectContainerEntry_fields(
        Legacy_StringObjectContainerEntry *stringObjectContainerEntry) {
    stringObjectContainerEntry->object =
            Legacy_ObjectComponentConstructorClassName(
                    "Legacy_StringObjectContainerEntry");

    static Constructable const constructable = {
            .constructor = (void *(*const)(void) )(
                    &Legacy_StringObjectContainerEntryConstructor)};
    stringObjectContainerEntry->object->constructable = &constructable;

    static Destructable const destructable = {
            .destructor = (void *(*const)(void *) )(
                    &Legacy_StringObjectContainerEntryDestructor)};
    stringObjectContainerEntry->object->destructable = &destructable;

    stringObjectContainerEntry->key   = NULL;
    stringObjectContainerEntry->value = NULL;
}

Legacy_StringObjectContainerEntry *
Legacy_StringObjectContainerEntryConstructor() {
    Legacy_StringObjectContainerEntry *instance = calloc(1, sizeof *instance);
    if (instance == NULL) { /* error handling here */
    }

    constructor_Legacy_StringObjectContainerEntry_fields(instance);

    return instance;
}

Legacy_StringObjectContainerEntry *
Legacy_StringObjectContainerEntryConstructorWithKeyAndValue(
        char *key, Legacy_Object *value) {
    Legacy_StringObjectContainerEntry *instance =
            Legacy_StringObjectContainerEntryConstructor();

    instance->key   = key;
    instance->value = value;

    return instance;
}

File: src/lib/oop/Object/Legacy_StringObjectContainerEntry.r
#ifndef LEGACY_STRINGOBJECTCONTAINERENTRY_H
#define LEGACY_STRINGOBJECTCONTAINERENTRY_H


#include "Legacy_Object.r"

/* Forward declaration of incomplete type */
typedef struct legacy_stringObjectContainerEntry
        Legacy_StringObjectContainerEntry;

struct legacy_stringObjectContainerEntry {

    /** `Legacy_StringObjectContainerEntry` implements `Legacy_ObjectComponent`. */
    Legacy_ObjectComponent *object;

    char *         key;
    Legacy_Object *value;
};


Legacy_StringObjectContainerEntry *
Legacy_StringObjectContainerEntryConstructor();

Legacy_StringObjectContainerEntry *
Legacy_StringObjectContainerEntryConstructorWithKeyAndValue(
        char *key, Legacy_Object *value);

#endif /* LEGACY_STRINGOBJECTCONTAINERENTRY_H */
